File: runtime/native/fs_module.go
package native

import (
	"io"
	"os"

	nruntime "novago/runtime"
)

func CreateFSModule() *nruntime.ObjectVal {
	fsProps := make(map[string]nruntime.RuntimeVal)

	fsProps["read"] = nruntime.MK_NATIVE_FN(func(args []nruntime.RuntimeVal, env *nruntime.Environment) nruntime.RuntimeVal {
		path := args[0].(*nruntime.StringVal).Value
		data, err := os.ReadFile(path)
		if err != nil {
			return nruntime.MK_NULL()
		}
		return nruntime.MK_STRING(string(data))
	})

	fsProps["write"] = nruntime.MK_NATIVE_FN(func(args []nruntime.RuntimeVal, env *nruntime.Environment) nruntime.RuntimeVal {
		path := args[0].(*nruntime.StringVal).Value
		data := args[1].(*nruntime.StringVal).Value
		err := os.WriteFile(path, []byte(data), 0644)
		if err != nil {
			return nruntime.MK_BOOL(false)
		}
		return nruntime.MK_BOOL(true)
	})

	fsProps["append"] = nruntime.MK_NATIVE_FN(func(args []nruntime.RuntimeVal, env *nruntime.Environment) nruntime.RuntimeVal {
		path := args[0].(*nruntime.StringVal).Value
		data := args[1].(*nruntime.StringVal).Value
		f, err := os.OpenFile(path, os.O_APPEND|os.O_CREATE|os.O_WRONLY, 0644)
		if err != nil {
			return nruntime.MK_BOOL(false)
		}
		defer f.Close()
		if _, err := f.WriteString(data); err != nil {
			return nruntime.MK_BOOL(false)
		}
		return nruntime.MK_BOOL(true)
	})

	fsProps["exists"] = nruntime.MK_NATIVE_FN(func(args []nruntime.RuntimeVal, env *nruntime.Environment) nruntime.RuntimeVal {
		path := args[0].(*nruntime.StringVal).Value
		_, err := os.Stat(path)
		if err == nil {
			return nruntime.MK_BOOL(true)
		}
		if os.IsNotExist(err) {
			return nruntime.MK_BOOL(false)
		}
		return nruntime.MK_BOOL(false)
	})

	fsProps["delete"] = nruntime.MK_NATIVE_FN(func(args []nruntime.RuntimeVal, env *nruntime.Environment) nruntime.RuntimeVal {
		path := args[0].(*nruntime.StringVal).Value
		err := os.RemoveAll(path)
		if err != nil {
			return nruntime.MK_BOOL(false)
		}
		return nruntime.MK_BOOL(true)
	})

	fsProps["mkdir"] = nruntime.MK_NATIVE_FN(func(args []nruntime.RuntimeVal, env *nruntime.Environment) nruntime.RuntimeVal {
		path := args[0].(*nruntime.StringVal).Value
		err := os.MkdirAll(path, 0755)
		if err != nil {
			return nruntime.MK_BOOL(false)
		}
		return nruntime.MK_BOOL(true)
	})

	fsProps["copy"] = nruntime.MK_NATIVE_FN(func(args []nruntime.RuntimeVal, env *nruntime.Environment) nruntime.RuntimeVal {
		src := args[0].(*nruntime.StringVal).Value
		dst := args[1].(*nruntime.StringVal).Value
		err := copyFile(src, dst)
		return nruntime.MK_BOOL(err == nil)
	})

	fsProps["move"] = nruntime.MK_NATIVE_FN(func(args []nruntime.RuntimeVal, env *nruntime.Environment) nruntime.RuntimeVal {
		src := args[0].(*nruntime.StringVal).Value
		dst := args[1].(*nruntime.StringVal).Value
		err := os.Rename(src, dst)
		return nruntime.MK_BOOL(err == nil)
	})

	fsProps["rename"] = nruntime.MK_NATIVE_FN(func(args []nruntime.RuntimeVal, env *nruntime.Environment) nruntime.RuntimeVal {
		old := args[0].(*nruntime.StringVal).Value
		new := args[1].(*nruntime.StringVal).Value
		err := os.Rename(old, new)
		return nruntime.MK_BOOL(err == nil)
	})

	fsProps["isFile"] = nruntime.MK_NATIVE_FN(func(args []nruntime.RuntimeVal, env *nruntime.Environment) nruntime.RuntimeVal {
		path := args[0].(*nruntime.StringVal).Value
		stat, err := os.Stat(path)
		if err != nil {
			return nruntime.MK_BOOL(false)
		}
		return nruntime.MK_BOOL(!stat.IsDir())
	})

	fsProps["isDir"] = nruntime.MK_NATIVE_FN(func(args []nruntime.RuntimeVal, env *nruntime.Environment) nruntime.RuntimeVal {
		path := args[0].(*nruntime.StringVal).Value
		stat, err := os.Stat(path)
		if err != nil {
			return nruntime.MK_BOOL(false)
		}
		return nruntime.MK_BOOL(stat.IsDir())
	})

	fsProps["list"] = nruntime.MK_NATIVE_FN(func(args []nruntime.RuntimeVal, env *nruntime.Environment) nruntime.RuntimeVal {
		dir := args[0].(*nruntime.StringVal).Value
		entries, err := os.ReadDir(dir)
		if err != nil {
			return nruntime.MK_ARRAY([]nruntime.RuntimeVal{})
		}
		var files []nruntime.RuntimeVal
		for _, e := range entries {
			files = append(files, nruntime.MK_STRING(e.Name()))
		}
		return nruntime.MK_ARRAY(files)
	})

	return nruntime.MK_OBJECT(fsProps)
}

func copyFile(src, dst string) error {
	in, err := os.Open(src)
	if err != nil {
		return err
	}
	defer in.Close()

	out, err := os.Create(dst)
	if err != nil {
		return err
	}
	defer out.Close()

	_, err = io.Copy(out, in)
	return err
}

File: novascript/src/runtime/native/fs_module.ts

import * as fs from "fs";
import * as path from "path";
import { RuntimeVal, MK_NATIVE_FN, MK_STRING, MK_BOOL, MK_NULL, MK_NUMBER, MK_OBJECT, MK_ARRAY } from "../values";

export function createFSModule() {
    const fsProps = new Map<string, RuntimeVal>();

    // Basic Read/Write
    fsProps.set("read", MK_NATIVE_FN((args) => {
        const p = (args[0] as any).value;
        if (!fs.existsSync(p)) return MK_NULL();
        return MK_STRING(fs.readFileSync(p, "utf-8"));
    }));

    fsProps.set("write", MK_NATIVE_FN((args) => {
        fs.writeFileSync((args[0] as any).value, (args[1] as any).value);
        return MK_BOOL(true);
    }));

    fsProps.set("append", MK_NATIVE_FN((args) => {
        fs.appendFileSync((args[0] as any).value, (args[1] as any).value);
        return MK_BOOL(true);
    }));

    // Advanced Operations
    fsProps.set("copy", MK_NATIVE_FN((args) => {
        const src = (args[0] as any).value;
        const dest = (args[1] as any).value;
        const recursive = (args[2] as any)?.value ?? true;
        fs.cpSync(src, dest, { recursive });
        return MK_BOOL(true);
    }));

    fsProps.set("move", MK_NATIVE_FN((args) => {
        const src = (args[0] as any).value;
        const dest = (args[1] as any).value;
        fs.renameSync(src, dest);
        return MK_BOOL(true);
    }));

    fsProps.set("rename", MK_NATIVE_FN((args) => {
        const oldPath = (args[0] as any).value;
        const newPath = (args[1] as any).value;
        fs.renameSync(oldPath, newPath);
        return MK_BOOL(true);
    }));

    fsProps.set("delete", MK_NATIVE_FN((args) => {
        const p = (args[0] as any).value;
        const recursive = (args[1] as any)?.value ?? true;
        fs.rmSync(p, { recursive, force: true });
        return MK_BOOL(true);
    }));

    fsProps.set("exists", MK_NATIVE_FN((args) => MK_BOOL(fs.existsSync((args[0] as any).value))));
    
    fsProps.set("mkdir", MK_NATIVE_FN((args) => {
        const p = (args[0] as any).value;
        fs.mkdirSync(p, { recursive: true });
        return MK_BOOL(true);
    }));

    fsProps.set("rmdir", MK_NATIVE_FN((args) => {
        const p = (args[0] as any).value;
        fs.rmSync(p, { recursive: true, force: true });
        return MK_BOOL(true);
    }));

    fsProps.set("isFile", MK_NATIVE_FN((args) => MK_BOOL(fs.statSync((args[0] as any).value).isFile())));
    fsProps.set("isDir", MK_NATIVE_FN((args) => MK_BOOL(fs.statSync((args[0] as any).value).isDirectory())));

    // List and Search
    fsProps.set("list", MK_NATIVE_FN((args) => {
        const dir = (args[0] as any).value;
        const files = fs.readdirSync(dir);
        return MK_ARRAY(files.map(f => MK_STRING(f)));
    }));

    fsProps.set("search", MK_NATIVE_FN((args) => {
        const pattern = (args[0] as any).value;
        const startDir = (args[1] as any)?.value || ".";
        const recursive = (args[2] as any)?.value ?? true;
        
        const results: string[] = [];
        const walk = (dir: string) => {
            const list = fs.readdirSync(dir);
            for (const file of list) {
                const fullPath = path.join(dir, file);
                if (file.includes(pattern)) results.push(fullPath);
                if (recursive && fs.statSync(fullPath).isDirectory()) {
                    walk(fullPath);
                }
            }
        };
        
        walk(startDir);
        return MK_ARRAY(results.map(r => MK_STRING(r)));
    }));

    return MK_OBJECT(fsProps);
}
